File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    int [] dr= {1,-1,0,0};
    int [] dc= {0,0,-1,1};


    class Pair{
        int row;
        int col;
        int time;
        public Pair(int row, int col, int time){
            this.row=row;
            this.col=col;
            this.time= time;
        }
    }


    public int orangesRotting(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int max_time=0;
        int fresh=0;
        Queue<Pair> qu= new LinkedList<>();
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(grid[i][j]==2){
                    qu.offer(new Pair(i, j, 0));
                }
                else if(grid[i][j]==1){
                    fresh++;
                }
            }
        }
        if (fresh==0){
            return 0;
        }
        while(!qu.isEmpty()){
            Pair curr= qu.poll();
            int crow= curr.row;
            int ccol= curr.col;
            int time= curr.time;
            max_time= Math.max(time, max_time);
            for(int k=0; k<4; k++){
                int nr = crow + dr[k];
                int nc = ccol + dc[k];
                if(nr>=0 && nc>=0 && nr<n && nc<m && grid[nr][nc]==1){
                    grid[nr][nc]=2;
                    fresh--;
                    qu.offer(new Pair(nr, nc, time+1));
                }
            }
        }
        return (fresh==0)?max_time:-1;
    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
from collections import deque

class Solution(object):

    class Pair:
        def __init__(self, row, col, time):
            self.row = row
            self.col = col
            self.time = time

    dr = [1, -1, 0, 0]
    dc = [0, 0, -1, 1]

    def orangesRotting(self, grid):

        n = len(grid)
        m = len(grid[0])

        fresh = 0
        max_time = 0

        qu = deque()
        for i in range(n):
            for j in range(m):

                if grid[i][j] == 2:
                    qu.append(self.Pair(i, j, 0))

                elif grid[i][j] == 1:
                    fresh += 1

        if fresh == 0:
            return 0

        while qu:

            p = qu.popleft()

            row = p.row
            col = p.col
            time = p.time

            max_time = max(max_time, time)

            for i in range(4):

                nr = row + self.dr[i]
                nc = col + self.dc[i]

                if (0 <= nr < n and
                    0 <= nc < m and
                    grid[nr][nc] == 1):

                    grid[nr][nc] = 2
                    fresh -= 1

                    qu.append(self.Pair(nr, nc, time + 1))

        if fresh > 0:
            return -1

        return max_time